tracer/src/Datadog.Trace/AspNet/SharedItems.cs

```diff
@@ -0,0 +1,67 @@
+// <copyright file="SharedItems.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#if NETFRAMEWORK
+
+using System.Collections.Generic;
+using System.Web;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.AspNet
+{
+    internal static class SharedItems
+    {
+        public const string HttpContextPropagatedResourceNameKey = "__Datadog.Trace.ClrProfiler.Managed.AspNetMvcIntegration-aspnet.resourcename";
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(SharedItems));
+
+        internal static void PushItem<T>(HttpContext context, string key, T item)
+        {
+            if (context == null)
+            {
+                return;
+            }
+
+            // Storing only the scope by default to avoid allocating a stack if no inner calls are done
+            if (!context.Items.Contains(key) || context.Items[key] is null)
+            {
+                context.Items[key] = item;
+            }
+            else if (context.Items[key] is Stack<T> stack)
+            {
+                stack.Push(item);
+            }
+            else if (context.Items[key] is T previousScope)
+            {
+                var newStack = new Stack<T>();
+                newStack.Push(previousScope);
+                newStack.Push(item);
+                context.Items[key] = newStack;
+            }
```

Comment:
minor, but  I don't think you need the initial `Contains`. I believe `context.Items[key]` returns `null` if it doesn't exist. Also, you probably only want to do the access once. 

```suggestion
            var existingItem = context.Items[key];
            if (existingItem is null)
            {
                context.Items[key] = item;
            }
            else if (existingItem is Stack<T> stack)
            {
                stack.Push(item);
            }
            else if (existingItem is T previousScope)
            {
                var newStack = new Stack<T>();
                newStack.Push(previousScope);
                newStack.Push(item);
                context.Items[key] = newStack;
            }
```

Reply:
Indeed. I was certain it threw so I haven't even checked. Thanks!